src/test/java/com/google/sps/SurveyServletTest.java

```diff
@@ -0,0 +1,268 @@
+import static org.junit.Assert.assertNull;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreOptions;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.IncompleteKey;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.PathElement;
+import com.google.cloud.datastore.testing.LocalDatastoreHelper;
+import java.io.IOException;
+import java.util.concurrent.TimeoutException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Tests for the SurveyServlet class.
+ *
+ * Make sure to set the project you're working on in your shell with:
+ * gcloud config set project [PROJECT_NAME]
+ *
+ * Skeleton code for Datastore emulator from:
+ * https://stackoverflow.com/questions/40348653/google-datastore-emulator-using-java-not-using-gae
+ */
+public class SurveyServletTest extends testData {
+
+    protected static LocalDatastoreHelper localDatastoreHelper;
+    protected Datastore datastore;
+    protected KeyFactory keyFactory;
+
+    @BeforeClass
+    public static void setUpClass() throws InterruptedException, IOException {
+        // Create and start a local datastore emulator on a random free port
+        System.out.println("[Datastore-Emulator] start");
+        localDatastoreHelper = LocalDatastoreHelper.create();
+        localDatastoreHelper.start();
+        System.out.println("[Datastore-Emulator] listening on port: " + localDatastoreHelper.getPort());
+
+        // Set the system property to tell the gcloud lib to use the datastore emulator
+        System.setProperty("DATASTORE_EMULATOR_HOST","localhost:" + localDatastoreHelper.getPort());
+    }
+
+    @Before
+    public void setUp() {
+        // Create the datastore instance
+        // Because of the system property set it in setUpClass() this
+        // datastore will be connected with the datastore emulator.
+        datastore = DatastoreOptions.getDefaultInstance().getService();
+        keyFactory = datastore.newKeyFactory().setProjectId("manage-at-scale-step-2020").setKind("TestEntity");
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        System.out.println("[Datastore-Emulator] reset");
+        // This resets the datastore after every test
+        localDatastoreHelper.reset();
+    }
+
+    @AfterClass
+    public static void tearDownClass() throws InterruptedException, IOException, TimeoutException {
+        System.out.println("[Datastore-Emulator] stop");
+        // This stops the datastore emulator after all tests are done
+        localDatastoreHelper.stop();
+    }
+
+
+    // Testing emulator properly resets between tests
+
+    @Test
+    public void testNotLeakingState1() {
```

Comment:
You probably don't need these tests, because we should assume that code we didn't write works correctly.

Reply:
Resolved in b544a05